### PR TITLE
async signing support

### PR DESCRIPTION
**Pull request by thehesiod**
This starts the large work of supporting https://github.com/aio-libs/aiobotocore/issues/619

Large changes:
- (api breaking) the result of`create_client` is now a required async context class
- `generate_presigned_url` is now an async call
- `Credentials.[access_key/secret_key/token]` now raise NotImplementedError because they won't call refresh like botocore.  Instead should use get_frozen_credentials async method

In general, this is going to increase our footprint significantly.

TODO
------
- [x] get tests/checks passing
- [x] ensured that all the `events.emit` calls have been properly wrapped (all the way up call chain)
- [x] add patches for new methods
- [x] port credential classes and callers, partial list: IMDSFetcher, ContainerMetadataFetcher, AssumeRoleWithWebIdentityCredentialFetcher, ProcessProvider, AssumeRoleCredentialFetcher
- [x] add tests for coverage (ports from botocore)
- [x] look into `botocore.utils.S3RegionRedirector.redirect_from_error`
- [x] adhoc test using real aws creds

### CODE REVIEW

**Comment by thehesiod** [comment]
with the changes as-is we support the simple testcase I list in https://github.com/aio-libs/aiobotocore/issues/658

**Comment by terrycain** [comment]
This'll help with #707 as we essentially want to make botocore.utils.S3RegionRedirector.redirect_from_error async

**Comment by thehesiod** [comment]
I'm going to try working on this this weekend

**Comment by terrycain** [comment]
@thehesiod hows it going?

**Comment by thehesiod** [comment]
@terrycain under a mountain of work with new baby almost here 🤯 :)

**Comment by terrycain** [comment]
Np man, congrats :smile:

**Comment by stalkerg** [comment]
I am very interesting in async generate_presigned_url, how I can help to merge this PR?

**Comment by thehesiod** [comment]
hello @stalkerg , you can take a  look at this PR.  We need to change `create_client` from being an async method to returning an async context class as mentioned in the description.  Also need to merge from master, and ensure all the emit methods that touch these changes are async.  In general need to do a lot of research to ensure this system will work, which means we probably need a lot of more testing.  This PR probably should be the place we do adaptation of the underlying botocore unittests.

**Comment by thehesiod** [comment]
not sure how this happened, but we're out of sync with botocore, I'm working on cleaning this up

**Comment by thehesiod** [comment]
omg, the whole credentials file needs to be ported!  I give up for awhile again, so much work!

**Comment by terrycain** [comment]
So I had a quick look at this just now - https://github.com/terrycain/aiobotocore/tree/streaming
Managed to get generate_presigned_url working... but as all the event emitting is async and the generate_presigned_url in the request signer is async, needed to copy the function and do some awaits, then replace the function that adds generate_presigned_url to the classes in the component tree thing.

Had a look into the credentials class, my god that file is nasty, literally everything needs to be patched as well as some classes from utils.py.

**Comment by thehesiod** [comment]
@terrycain welcome to hell lol

**Comment by terrycain** [comment]
You mind moving your branch to the aiobotocore repo, then I could add to it, if you don't mind that is :D

**Comment by thehesiod** [comment]
@terrycain the branch is here, it's just called `thehesiod/streaming`.  I don't have a fork of this repo

**Comment by terrycain** [comment]
Duh, i even grabbed your changes from it earlier, my brain is not working today. Yeah I'll merge in my fixes to signatures, see where we're at and possibly start on that horrific credentials file

**Comment by terrycain** [comment]
Ok so from my basic test of s3.list_buckets, it seems to be using the asyncified credentials and signatures. Needs many many many tests but we're getting there.

**Comment by thehesiod** [comment]
i say when this and the 3.8 PR land we actually bump to 1.0, what do you think?

**Comment by terrycain** [comment]
And the aiomoto PR, by then when we hit 1.0 we'll have a pretty decent test suite with hopefully a good amount of coverage.

**Comment by thehesiod** [comment]
merged in my py38 changes

File: aiobotocore/session.py [hunk 1]
```diff
@@ -1,25 +1,80 @@
...
-from .client import AioClientCreator
+from .client import AioClientCreator, AioBaseClient
+from .hooks import AioHierarchicalEmitter, AioEventAliaser
 from .parsers import AioResponseParserFactory
+from .signers import add_generate_presigned_url
+from .credentials import create_credential_resolver, AioCredentials
 
 
-class AioSession(botocore.session.Session):
+class ClientCreatorContext:
```
Comment (thehesiod) [inline aiobotocore/session.py]: I'm sure this is available from python somewhere

**Comment by thehesiod** [comment]
btw for testing we should try to port the botocore tests as much as possible.  We can probably simplify to just one test per area we're replacing since we're building on top of a tested Botocore.

**Comment by terrycain** [comment]
Yeah, I've had a quick look over the credentials tests mainly the parts where we've changed the requests out for aiohttp. For the tests, they're all unittests, would it be worth grabbing a load and converting to pytest if its simple?

**Comment by thehesiod** [comment]
Botocore uses nose I think, perhaps we could have a separate set of tests that are translated botocore tests. I was wondering if we run the tests as similar as possible where the test runner keeps on running the test while it returns an awaitable perhaps we could almost run the tests unchanged

**Comment by terrycain** [comment]
So I had a look over their credentials tests, for the parts we've patched I think we can get away with stealing one or two tests per class. Not finished on the credentials tests yet but will see where the coverage leaves us

**Comment by terrycain** [comment]
We're changing generate_presign_url to async. We'll need to bump to a major version, but in the meantime would it make sense to add a warning to it? So people are aware when they update

**Comment by terrycain** [comment]
:) coverage is getting there

**Comment by terrycain** [comment]
So the S3RegionRedirector works, there is an edge case where it would not, but I cant see a case where S3 won't return the redirected region header in the redirect response so am not patching the rest of that for now. Signer now respects changes in regions when signing retried requests.

**Comment by thehesiod** [comment]
> We're changing generate_presign_url to async. We'll need to bump to a major version, but in the meantime would it make sense to add a warning to it? So people are aware when they update

Eh. Since it's a major change in don't think we need to. Maybe in front page we can add upcoming changes to expect

**Comment by thehesiod** [comment]
Sorry for not helping recently. Have a major project we're trying to got through at work

**Comment by terrycain** [comment]
Ok so I think the test coverage is better than it was pre-pr :wink: I'm going to test the code against real creds including the assume role variants to see if they actually work for real.

@thehesiod is there anything glaringly obvious were missing apart from spamming patch hashes for stuff we've overridden?

Then we'll merge in any changes from master (which will probably mean updating test patches again, but worthwhile) and then I think we're good

**Comment by terrycain** [comment]
Also now supports the new v2 instance metadata endpoint which requires you doing a PUT request to get a session token of sorts (https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html)

**Comment by terrycain** [comment]
Merged in master, no changes required.

Tested with the following env vars
```
AWS_ROLE_ARN=arn:aws:iam::100000000000:role/test
AWS_WEB_IDENTITY_TOKEN_FILE=/tmp/jwt
AWS_ROLE_SESSION_NAME=session1
```
And script:
```python
import aiobotocore
import asyncio
async def main():
    session = aiobotocore.session.AioSession()
    async with session.create_client('s3') as client:
        print(await client.get_object(Bucket='terry', Key='test'))
asyncio.get_event_loop().run_until_complete(main())
```
It complained because my JWT file was not a JWT but it seemed to call assume role with web identity fine, and that's probably the most complicated path through the credentials mess. I believe it'll work with the AWS EKS Kube IAM role integration stuff (https://aws.amazon.com/blogs/opensource/introducing-fine-grained-iam-roles-service-accounts/).

@thehesiod when you get some time, you mind reviewing this and checking the emitter stuff (think we're all there)

The hell is coming to an end :D

**Comment by thehesiod** [comment]
Hey Terry I'll get to this asap, having a hard time at home with the two kids and work project. Really appreciate all the help!

**Comment by terrycain** [comment]
Npnp when you can :smile:

File: tests/test_patches.py [hunk 2]
```diff
@@ -110,15 +217,56 @@
...
     Waiter.wait: {'5502a89ed740fb5d6238a6f72a3a08efc1a9f43b'},
     create_waiter_with_client: {'c3d12c9a4293105cc8c2ecfc7e69a2152ad564de'},
 }
 
+# REPLACE = backwards incompatible change
+# APPEND = officially supporting more versions of botocore/aiohttp
+
+# If you're changing these, most likely need to update setup.py as well.
+
 _PROTOCOL_PARSER_CONTENT = {'ec2', 'query', 'json', 'rest-json', 'rest-xml'}
```
Comment (thehesiod) [inline tests/test_patches.py]: @terrycain wow =)

Comment (terrycain) [reply tests/test_patches.py]: Yeah... the rabbit hole ran deep on this one. Mainly the credentials class and the utils that it used.

**Comment by thehesiod** [comment]
ok tonight I'll finish reviewing this because i need it for a project :)

File: aiobotocore/credentials.py [hunk 3]
```diff
@@ -0,0 +1,774 @@
...
+            obj._access_key, obj._secret_key,
+            obj._token, obj._expiry_time,
+            obj._refresh_using, obj.method,
+            obj._time_fetcher
+        )
+
+    # Redeclaring the properties so it doesnt call refresh
+    # Have to redeclare setter as we're overriding the getter
+    @property
+    def access_key(self):
```
Comment (thehesiod) [inline aiobotocore/credentials.py]: @terrycain this is a problem no?  Look at `HmacV1PostAuth.add_auth` : `fields['AWSAccessKeyId'] = self.credentials.access_key`

and `HmacV1QueryAuth._inject_signature` : `query_dict['AWSAccessKeyId'] = self.credentials.access_key`. there are a few more places too

Comment (thehesiod) [reply aiobotocore/credentials.py]: this is kinda the lynchpin :)

**Comment by thehesiod** [comment]
@terrycain this is definitely hero status work :)

**Comment by thehesiod** [comment]
@terrycain ok finished, everything looks great except for the access_key etc properties, I made them raise for now because they won't work since they're not calling refresh.  We need to figure out what to do with these and then we can merge!

**Comment by thehesiod** [comment]
btw I re-organized the code so that it would match the order in botocore to make diffs easier

**Comment by thehesiod** [comment]
ok, I think we're good in terms of emitters, now to fix tests again

**Comment by thehesiod** [comment]
ok fixed tests, and set `access_key`+ `secret_key` + `token` properties to raise NotImplementedError, instead callers should use `get_frozen_credentials`. thoughts?

**Comment by terrycain** [comment]
Ok so according to something I read in the botocore source/docs somewhere, everything is supposed to call get_frozen_credentials, lemme have a look

File: aiobotocore/credentials.py (hunk 3 above)
Comment (terrycain) [reply aiobotocore/credentials.py]: Yeah true

Comment (terrycain) [reply aiobotocore/credentials.py]: Though technically all signing auth classes are passed frozen credentials :D but yeah, the only safe place to read access_key... would be from the ReadOnlyCredential class retrieved from get_frozen_credentials

**Comment by terrycain** [comment]
Yeah I think were good. So tests are failing because of an old pyyaml version having a CVE, pretty sure aws-cli is the culprit as they seem to have moved to using ruamel.yaml. 

Should we just do the usual bump botocore version stuff here and then merge?

**Comment by terrycain** [comment]
Also tests run fine for me so the access_key changes are good

**Comment by thehesiod** [comment]
ya lets bump and merge!  We're hitting this CVE at work too so bump would be awesome, if you don't get to in in a few hours I'll do it and merge. 🎆 🍾

**Comment by thehesiod** [comment]
ok doing it now

**Comment by terrycain** [comment]
Yay :smile:

**Comment by thehesiod** [comment]
@terrycain thanks so much for your help!  Wouldn't have been possible without you, at least for a long time :]

**Comment by terrycain** [comment]
No problem ;-). Just realized we need to override session._set_credentials() as otherwise it breaks when access and secret key are passed explicitly :(. Will pr in morning if you don't get there first, am too tired now :D

**Comment by thehesiod** [comment]
Hmm I left the set, just not the get. Too tired to look now too though :)

**Pull request merged**